### PR TITLE
Modern cmake targets

### PR DESCRIPTION
**Pull request by lettleli**
Split the CMakeLists.txt file into one main CMakeLists.txt and additional CMakeLists.txt files for the testing and library subdirectories. Use "modern" CMake with only targets and target properties, which are also exported into the build and install tree.

Keep the output and build behaviour of the old cmake build process.

Use ctest in the coverage script.

### CODE REVIEW

**Comment by pdebuyl** [comment]
Hi @lettleli 

Thank you for your interest and your PR!

I have started to browse through the changes and have some questions:
- Is it necessary to use cmake as recent as 3.10 ? Travis, which I use for continuous integrations has only 3.9. Many Fortran users use HPC systems that don't have up to date programming tools.
- Do you have a link to a page or documentation that explains a bit the benefits of using the modern features such as "targets"? Programs are target anyway no in cmake?
- What does `GNUInstallDirs` do?
- Is there any reason for the `get_filename_component` in `test/CMakeLists.txt`? The name is directly the loop variable here.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,20 +1,83 @@
...
+    INSTALL_DESTINATION ${config_install_dir}
+)
+install(FILES ${cmake_files_dir}/${project_config}
+    DESTINATION ${config_install_dir}
+)
+
+#create package config for build tree
+configure_package_config_file(cmake/PackageConfig.cmake.in ${config_build_dir}/${project_config}
+    INSTALL_DESTINATION ${config_build_dir}
+    INSTALL_PREFIX ${PROJECT_BINARY_DIR}
```
Comment (lettleli) [inline CMakeLists.txt:82]: INSTALL_PREFIX of configure_package_config_file is the only feature not supported by CMake 3.0.
it is not actually relevant until now, but removing it, would mean that one shouldn't use this function to create the package config file for the build tree.

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,16 @@
+foreach(test_src test_tester_1.f90
+                 test_tester_2.f90
+                 test_tester_3.f90
+                 test_tester_4.f90
+                 test_tester_5.f90
+                 test_tester_6.f90)
+    get_filename_component(test_name ${test_src} NAME_WE)
```
Comment (lettleli) [inline test/CMakeLists.txt:7]: _get_filename_component_ with NAME_WE cuts off the extension, in this case it is always the same extension, but if you have extensions like f, F, f90, F90,... you can't add the extension in the loop like it was done before.

**Review by lettleli — COMMENTED** [review]
Hi @pdebuyl 

- The minimum required Version I got it to build today was 3.1.0, "modern" CMake is at least 2.8.12, better 3.0.
- Then main difference between the "old" and "modern" CMake style should be, that you don't use global functions like _include_directories_ any more. These informations are now set target specific and every target has private and interface properities and only the interface properties are inherited by other targets which depend on them. When using external CMake projects which export their targets it is then sufficient to only let CMake find the package config file and link the needed target, all informations to use a certain target are already included in its properties.
See for example:
  - https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd
  - https://cliutils.gitlab.io/modern-cmake/
- _GNUInstallDirs_ defines the CMAKE_INSTALL_... variables according to the GNU standard.
See for example: https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html?highlight=gnuinstalldirs

**Comment by pdebuyl** [comment]
Thank you for the replies @lettleli !

I have started to look at the changes and at the links you gave. For safety I will check that the code works from my Fortran projects that rely on it. I'll also take advantage of your PR to merge the branches I had left on the side for CI integration and coverage analysis.

Cheers,

Pierre

PS: I should be able to get to this by the end of next week.

**Comment by pdebuyl** [comment]
Thank you, merged!

**Pull request merged**